### ISSUE TITLE
(wrong-type-argument stringp nil) when typing additional letters in an Python argument

### ISSUE THREAD

**Issue by bgoodr** [jorgenschaefer/elpy#1110]
I am getting errors when simply typing in additional characters in an argument value of a Python function in a elpy-enabled Python buffer. The error is:

    (wrong-type-argument stringp nil)

Steps to reproduce:

Enable debugging on error:

    M-x toggle-debug-on-error

Add the following to /tmp/xxx.py by pasting it into an Emacs buffer:

    """Example module."""
    
    
    def do_something_important(x):
        print x

Position the point after the first `x` (in the argument list of the `do_something_important` function).

Start typing additional `x`'s. At about the third additional `x`, the elpy code throws an error which is captured as follows:

    Debugger entered--Lisp error: (wrong-type-argument stringp nil)
      signal(wrong-type-argument (stringp nil))
      (if (and (consp error-object) (not (consp (car error-object)))) (signal (car error-object) (cdr error-object)) (let ((message (cdr (assq (quote message) error-object))) (code (cdr (assq (quote code) error-object))) (data (cdr (assq (quote data) error-object)))) (cond ((not (numberp code)) (error "Bad response from RPC: %S" error-object)) ((< code 300) (message "Elpy warning: %s" message)) ((< code 500) (error "Elpy error: %s" message)) ((and elpy-rpc-error-timeout elpy-rpc--last-error-popup (<= (float-time) (+ elpy-rpc--last-error-popup elpy-rpc-error-timeout))) (message "Elpy error popup ignored, see `elpy-rpc-error-timeout': %s" message)) ((not elpy-disable-backend-error-display) (let ((config (elpy-config--get-config))) (progn (set-marker help-window-point-marker nil) (let (... ...) (let* ... ... ... ...))) (setq elpy-rpc--last-error-popup (float-time)))))))
      elpy-rpc--default-error-callback((wrong-type-argument stringp nil))
      funcall(elpy-rpc--default-error-callback (wrong-type-argument stringp nil))
      (save-current-buffer (set-buffer (elpy-promise-buffer promise)) (funcall error-callback reason))
      (if (buffer-live-p (elpy-promise-buffer promise)) (save-current-buffer (set-buffer (elpy-promise-buffer promise)) (funcall error-callback reason)) (let ((temp-buffer (generate-new-buffer " *temp*"))) (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn (funcall error-callback reason)) (and (buffer-name temp-buffer) (kill-buffer temp-buffer))))))
      (progn (if (buffer-live-p (elpy-promise-buffer promise)) (save-current-buffer (set-buffer (elpy-promise-buffer promise)) (funcall error-callback reason)) (let ((temp-buffer (generate-new-buffer " *temp*"))) (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn (funcall error-callback reason)) (and (buffer-name temp-buffer) (kill-buffer temp-buffer)))))))
      (if error-callback (progn (if (buffer-live-p (elpy-promise-buffer promise)) (save-current-buffer (set-buffer (elpy-promise-buffer promise)) (funcall error-callback reason)) (let ((temp-buffer (generate-new-buffer " *temp*"))) (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn (funcall error-callback reason)) (and (buffer-name temp-buffer) (kill-buffer temp-buffer))))))))
      (let ((error-callback (elpy-promise-error-callback promise))) (if error-callback (progn (if (buffer-live-p (elpy-promise-buffer promise)) (save-current-buffer (set-buffer (elpy-promise-buffer promise)) (funcall error-callback reason)) (let ((temp-buffer (generate-new-buffer " *temp*"))) (save-current-buffer (set-buffer temp-buffer) (unwind-protect (progn ...) (and ... ...))))))))
      (unwind-protect (let ((error-callback (elpy-promise-error-callback promise))) (if error-callback (progn (if (buffer-live-p (elpy-promise-buffer promise)) (save-current-buffer (set-buffer (elpy-promise-buffer promise)) (funcall error-callback reason)) (let ((temp-buffer ...)) (save-current-buffer (set-buffer temp-buffer) (unwind-protect ... ...))))))) (elpy-promise-set-resolved promise))
      (progn (unwind-protect (let ((error-callback (elpy-promise-error-callback promise))) (if error-callback (progn (if (buffer-live-p (elpy-promise-buffer promise)) (save-current-buffer (set-buffer ...) (funcall error-callback reason)) (let (...) (save-current-buffer ... ...)))))) (elpy-promise-set-resolved promise)))
      (if (not (elpy-promise-resolved-p promise)) (progn (unwind-protect (let ((error-callback (elpy-promise-error-callback promise))) (if error-callback (progn (if (buffer-live-p ...) (save-current-buffer ... ...) (let ... ...))))) (elpy-promise-set-resolved promise))))
      elpy-promise-reject([*elpy-promise* (closure ((callback . #[257 "\211\203\301\242\307=\204\301\240\207\306\f\300\310!BC
\311\312DC\216\313\302\303\304\305$)\207" [#("xxxx" 0 4 (fontified t)) (done) #<buffer xxx.py> #<window 8 on xxx.py> 143 56 elpy-company-backend done company--preprocess-candidates funcall #[0 "?\205\n\303\211\211\207" [company-candidates company-backend company-candidates-cache nil] 3 "\n\n(fn)"] company-idle-begin company-backend company-candidates-cache] 6 "\n\n(fn CANDIDATES)"]) (ignored) (arg . #("xxxx" 0 4 (fontified t))) (command . candidates) t) (result) (elpy-company--cache-clear) (funcall callback (cond (result (elpy-company--cache-completions arg result)) ((> (length arg) company-minimum-prefix-length) (elpy--sort-and-strip-duplicates (company-dabbrev-code (quote candidates) arg))) (t nil)))) elpy-rpc--default-error-callback #<buffer xxx.py> nil] (wrong-type-argument stringp nil))
      (condition-case err (save-current-buffer (set-buffer (elpy-promise-buffer promise)) (funcall success-callback value)) (error (elpy-promise-reject promise err)))
      (progn (condition-case err (save-current-buffer (set-buffer (elpy-promise-buffer promise)) (funcall success-callback value)) (error (elpy-promise-reject promise err))))
      (if success-callback (progn (condition-case err (save-current-buffer (set-buffer (elpy-promise-buffer promise)) (funcall success-callback value)) (error (elpy-promise-reject promise err)))))
      (let ((success-callback (elpy-promise-success-callback promise))) (if success-callback (progn (condition-case err (save-current-buffer (set-buffer (elpy-promise-buffer promise)) (funcall success-callback value)) (error (elpy-promise-reject promise err))))))
      (unwind-protect (let ((success-callback (elpy-promise-success-callback promise))) (if success-callback (progn (condition-case err (save-current-buffer (set-buffer (elpy-promise-buffer promise)) (funcall success-callback value)) (error (elpy-promise-reject promise err)))))) (elpy-promise-set-resolved promise))
      (progn (unwind-protect (let ((success-callback (elpy-promise-success-callback promise))) (if success-callback (progn (condition-case err (save-current-buffer (set-buffer ...) (funcall success-callback value)) (error (elpy-promise-reject promise err)))))) (elpy-promise-set-resolved promise)))
      (if (not (elpy-promise-resolved-p promise)) (progn (unwind-protect (let ((success-callback (elpy-promise-success-callback promise))) (if success-callback (progn (condition-case err (save-current-buffer ... ...) (error ...))))) (elpy-promise-set-resolved promise))))
      elpy-promise-resolve([*elpy-promise* (closure ((callback . #[257 "\211\203\301\242\307=\204\301\240\207\306\f\300\310!BC
\311\312DC\216\313\302\303\304\305$)\207" [#("xxxx" 0 4 (fontified t)) (done) #<buffer xxx.py> #<window 8 on xxx.py> 143 56 elpy-company-backend done company--preprocess-candidates funcall #[0 "?\205\n\303\211\211\207" [company-candidates company-backend company-candidates-cache nil] 3 "\n\n(fn)"] company-idle-begin company-backend company-candidates-cache] 6 "\n\n(fn CANDIDATES)"]) (ignored) (arg . #("xxxx" 0 4 (fontified t))) (command . candidates) t) (result) (elpy-company--cache-clear) (funcall callback (cond (result (elpy-company--cache-completions arg result)) ((> (length arg) company-minimum-prefix-length) (elpy--sort-and-strip-duplicates (company-dabbrev-code (quote candidates) arg))) (t nil)))) elpy-rpc--default-error-callback #<buffer xxx.py> nil] nil)
      (if error-object (elpy-promise-reject promise error-object) (elpy-promise-resolve promise result))
      (let ((promise (gethash call-id elpy-rpc--backend-callbacks))) (if (not promise) (progn (error "Received a response for unknown call-id %s" call-id))) (remhash call-id elpy-rpc--backend-callbacks) (if error-object (elpy-promise-reject promise error-object) (elpy-promise-resolve promise result)))
      (let ((call-id (cdr (assq (quote id) json))) (error-object (cdr (assq (quote error) json))) (result (cdr (assq (quote result) json)))) (let ((promise (gethash call-id elpy-rpc--backend-callbacks))) (if (not promise) (progn (error "Received a response for unknown call-id %s" call-id))) (remhash call-id elpy-rpc--backend-callbacks) (if error-object (elpy-promise-reject promise error-object) (elpy-promise-resolve promise result))))
      elpy-rpc--handle-json(((id . 22) (result)))
      (cond (did-read-json (delete-region (point-min) line-end) (elpy-rpc--handle-json json)) ((looking-at "elpy-rpc ready\n") (replace-match "") (elpy-rpc--check-backend-version "1.1")) ((looking-at "elpy-rpc ready (\\([^ ]*\\))\n") (let ((rpc-version (match-string 1))) (replace-match "") (elpy-rpc--check-backend-version rpc-version))) ((looking-at ".*No module named elpy\n") (replace-match "") (elpy-config-error "Elpy module not found")) (t (let ((line (buffer-substring (point-min) line-end))) (delete-region (point-min) line-end) (elpy-rpc--handle-unexpected-line line))))
      (let ((line-end (point)) (json nil) (did-read-json nil)) (goto-char (point-min)) (condition-case err (progn (setq json (let ((json-array-type (quote list))) (json-read))) (if (listp json) (setq line-end (1+ (point)) did-read-json t) (goto-char (point-min)))) (error (goto-char (point-min)))) (cond (did-read-json (delete-region (point-min) line-end) (elpy-rpc--handle-json json)) ((looking-at "elpy-rpc ready\n") (replace-match "") (elpy-rpc--check-backend-version "1.1")) ((looking-at "elpy-rpc ready (\\([^ ]*\\))\n") (let ((rpc-version (match-string 1))) (replace-match "") (elpy-rpc--check-backend-version rpc-version))) ((looking-at ".*No module named elpy\n") (replace-match "") (elpy-config-error "Elpy module not found")) (t (let ((line (buffer-substring (point-min) line-end))) (delete-region (point-min) line-end) (elpy-rpc--handle-unexpected-line line)))))
      (while (progn (goto-char (point-min)) (search-forward "\n" nil t)) (let ((line-end (point)) (json nil) (did-read-json nil)) (goto-char (point-min)) (condition-case err (progn (setq json (let ((json-array-type ...)) (json-read))) (if (listp json) (setq line-end (1+ (point)) did-read-json t) (goto-char (point-min)))) (error (goto-char (point-min)))) (cond (did-read-json (delete-region (point-min) line-end) (elpy-rpc--handle-json json)) ((looking-at "elpy-rpc ready\n") (replace-match "") (elpy-rpc--check-backend-version "1.1")) ((looking-at "elpy-rpc ready (\\([^ ]*\\))\n") (let ((rpc-version (match-string 1))) (replace-match "") (elpy-rpc--check-backend-version rpc-version))) ((looking-at ".*No module named elpy\n") (replace-match "") (elpy-config-error "Elpy module not found")) (t (let ((line (buffer-substring ... line-end))) (delete-region (point-min) line-end) (elpy-rpc--handle-unexpected-line line))))))
      (save-current-buffer (set-buffer buffer) (goto-char (point-max)) (let ((pre-command-hook nil)) (save-excursion (save-restriction (let ((G14036 (format "elpy-rpc--filter output %S" output))) (save-current-buffer (set-buffer (get-buffer-create bg-debug-buffer-name)) (goto-char (point-max)) (insert G14036) (insert "\n")))))) (insert output) (while (progn (goto-char (point-min)) (search-forward "\n" nil t)) (let ((line-end (point)) (json nil) (did-read-json nil)) (goto-char (point-min)) (condition-case err (progn (setq json (let (...) (json-read))) (if (listp json) (setq line-end (1+ ...) did-read-json t) (goto-char (point-min)))) (error (goto-char (point-min)))) (cond (did-read-json (delete-region (point-min) line-end) (elpy-rpc--handle-json json)) ((looking-at "elpy-rpc ready\n") (replace-match "") (elpy-rpc--check-backend-version "1.1")) ((looking-at "elpy-rpc ready (\\([^ ]*\\))\n") (let ((rpc-version ...)) (replace-match "") (elpy-rpc--check-backend-version rpc-version))) ((looking-at ".*No module named elpy\n") (replace-match "") (elpy-config-error "Elpy module not found")) (t (let ((line ...)) (delete-region (point-min) line-end) (elpy-rpc--handle-unexpected-line line)))))))
      (progn (save-current-buffer (set-buffer buffer) (goto-char (point-max)) (let ((pre-command-hook nil)) (save-excursion (save-restriction (let ((G14036 ...)) (save-current-buffer (set-buffer ...) (goto-char ...) (insert G14036) (insert "\n")))))) (insert output) (while (progn (goto-char (point-min)) (search-forward "\n" nil t)) (let ((line-end (point)) (json nil) (did-read-json nil)) (goto-char (point-min)) (condition-case err (progn (setq json (let ... ...)) (if (listp json) (setq line-end ... did-read-json t) (goto-char ...))) (error (goto-char (point-min)))) (cond (did-read-json (delete-region (point-min) line-end) (elpy-rpc--handle-json json)) ((looking-at "elpy-rpc ready\n") (replace-match "") (elpy-rpc--check-backend-version "1.1")) ((looking-at "elpy-rpc ready (\\([^ ]*\\))\n") (let (...) (replace-match "") (elpy-rpc--check-backend-version rpc-version))) ((looking-at ".*No module named elpy\n") (replace-match "") (elpy-config-error "Elpy module not found")) (t (let (...) (delete-region ... line-end) (elpy-rpc--handle-unexpected-line line))))))))
      (if (and buffer (buffer-live-p buffer)) (progn (save-current-buffer (set-buffer buffer) (goto-char (point-max)) (let ((pre-command-hook nil)) (save-excursion (save-restriction (let (...) (save-current-buffer ... ... ... ...))))) (insert output) (while (progn (goto-char (point-min)) (search-forward "\n" nil t)) (let ((line-end (point)) (json nil) (did-read-json nil)) (goto-char (point-min)) (condition-case err (progn (setq json ...) (if ... ... ...)) (error (goto-char ...))) (cond (did-read-json (delete-region ... line-end) (elpy-rpc--handle-json json)) ((looking-at "elpy-rpc ready\n") (replace-match "") (elpy-rpc--check-backend-version "1.1")) ((looking-at "elpy-rpc ready (\\([^ ]*\\))\n") (let ... ... ...)) ((looking-at ".*No module named elpy\n") (replace-match "") (elpy-config-error "Elpy module not found")) (t (let ... ... ...))))))))
      (let ((buffer (process-buffer process))) (if (and buffer (buffer-live-p buffer)) (progn (save-current-buffer (set-buffer buffer) (goto-char (point-max)) (let ((pre-command-hook nil)) (save-excursion (save-restriction (let ... ...)))) (insert output) (while (progn (goto-char (point-min)) (search-forward "\n" nil t)) (let ((line-end ...) (json nil) (did-read-json nil)) (goto-char (point-min)) (condition-case err (progn ... ...) (error ...)) (cond (did-read-json ... ...) (... ... ...) (... ...) (... ... ...) (t ...))))))))
      elpy-rpc--filter(#<process  *elpy-rpc [project:/tmp/ python:/home/brentg/python_virtualenv/bin/python]*> "{\"result\": [], \"id\": 22}\n")
    
`elpy-version` is "1.14.1"

`pip list` emits:

    autopep8 (1.3.1)
    configparser (3.5.0)
    enum34 (1.1.6)
    flake8 (3.3.0)
    google-api-python-client (1.5.5)
    httplib2 (0.9.2)
    importmagic (0.1.7)
    jedi (0.10.0)
    lxml (3.7.3)
    mccabe (0.6.1)
    oauth2client (4.0.0)
    pip (9.0.1)
    pyasn1 (0.1.9)
    pyasn1-modules (0.0.8)
    pycodestyle (2.3.1)
    pyflakes (1.5.0)
    requests (2.10.0)
    rsa (3.4.2)
    selenium (3.0.1)
    setuptools (24.0.1)
    six (1.10.0)
    svgutils (0.2.0)
    Unidecode (0.4.19)
    uritemplate (3.0.0)
    wheel (0.29.0)

python version:

    Python 2.7.11+ (default, Apr 17 2016, 14:00:29) 
    [GCC 5.3.1 20160413] on linux2

If I need to upgrade pip packages or elpy itself, I will do so.

Thanks,
Brent

**Comment by jorgenschaefer** [comment]
Hello, and thanks for the report! This looks a lot like #1101 – could try to remove all .elc files from your .emacs.d directory and recompile those files?

**Comment by bgoodr** [comment]
Summary:
========

What you suggested works, but now I want to know what
caused this. Bug in package.el? User error on my part?

TL;DR Detail:
=============

Before altered any files inside `~/.emacs.d`, I've saved it off to a
`.tar.gz` file, just in case we need to back out of it. Glad I did
because it took me several tries to script byte-compiling all of the
Elisp files as the way I interpreted should be done. There is probably
a one-liner equivalent of what I did below.  :-( 

So, running:

    cd ~/.emacs.d
    find . -name '*.elc'

gives:

    ./elpa/elpy-20170303.621/elpy-refactor.elc
    ./elpa/elpy-20170303.621/elpy-django.elc
    ./elpa/elpy-20170303.621/elpy-profile.elc
    ./elpa/elpy-20170303.621/snippets/python-mode/.yas-setup.elc
    ./elpa/highlight-indentation-20161012.209/highlight-indentation.elc
    ./elpa/highlight-indentation-20161012.209/highlight-indentation-pkg.elc
    ./elpa/pyvenv-20170224.538/pyvenv.elc
    ./elpa/pyvenv-20170224.538/pyvenv-pkg.elc
    ./elpa/s-20160928.636/s.elc
    ./elpa/s-20160928.636/s-pkg.elc
    ./elpa/ivy-20170321.1047/ivy.elc
    ./elpa/ivy-20170321.1047/ivy-overlay.elc
    ./elpa/ivy-20170321.1047/colir.elc
    ./elpa/company-20170315.1754/company-etags.elc
    ./elpa/company-20170315.1754/company-dabbrev-code.elc
    ./elpa/company-20170315.1754/company-bbdb.elc
    ./elpa/company-20170315.1754/company-elisp.elc
    ./elpa/company-20170315.1754/company-oddmuse.elc
    ./elpa/company-20170315.1754/company-gtags.elc
    ./elpa/company-20170315.1754/company.elc
    ./elpa/company-20170315.1754/company-abbrev.elc
    ./elpa/company-20170315.1754/company-template.elc
    ./elpa/company-20170315.1754/company-xcode.elc
    ./elpa/company-20170315.1754/company-eclim.elc
    ./elpa/company-20170315.1754/company-semantic.elc
    ./elpa/company-20170315.1754/company-cmake.elc
    ./elpa/company-20170315.1754/company-keywords.elc
    ./elpa/company-20170315.1754/company-files.elc
    ./elpa/company-20170315.1754/company-capf.elc
    ./elpa/company-20170315.1754/company-ispell.elc
    ./elpa/company-20170315.1754/company-dabbrev.elc
    ./elpa/company-20170315.1754/company-clang.elc
    ./elpa/company-20170315.1754/company-nxml.elc
    ./elpa/company-20170315.1754/company-tempo.elc
    ./elpa/company-20170315.1754/company-yasnippet.elc
    ./elpa/company-20170315.1754/company-css.elc
    ./elpa/yasnippet-20170310.1724/yasnippet.elc
    ./elpa/yasnippet-20170310.1724/snippets/python-mode/.yas-setup.elc
    ./elpa/yasnippet-20170310.1724/snippets/c++-mode/.yas-setup.elc
    ./elpa/yasnippet-20170310.1724/snippets/prog-mode/.yas-setup.elc
    ./elpa/find-file-in-project-20161202.2205/find-file-in-project-pkg.elc
    ./elpa/find-file-in-project-20161202.2205/find-file-in-project.elc

I interpreted your instructions to mean to delete all `.elc` files in
each directory that contained at least one `.elc` file, and then
recompile all `.el` files in that directory (yes, yielding more `.elc`
files than what we started with, which I assumed was ok at this point).

I then ran this script:

    #!/bin/bash

    cd ~/.emacs.d
    dirs=$(find . -type f | grep '\.elc$' | xargs -n1 dirname | sort | uniq)
    for dir in $dirs
    do
      echo
      echo "Processing directory $dir:"
      (
        cd $dir

        echo
        echo ".elc fileset before recompiling:"
        elc_files=$(find . -maxdepth 1 -type f | grep '\.elc$')
        echo "$elc_files" | xargs -n1000 ls -ld

        echo
        el_files=$(find . -maxdepth 1 -type f | grep '\.el$')
        echo "Recompiling:"
        rm -f $elc_files
        emacs -batch -f batch-byte-compile $el_files

        echo
        echo ".elc fileset after recompiling:"
        elc_files=$(find . -maxdepth 1 -type f | grep '\.elc$')
        echo "$elc_files" | xargs -n1000 ls -ld

      ) 2>&1 | sed 's%^%  %g'
    done

And produced a lot of output which I will attached for future
reference since there are errors that we probably can ignore due to
the `load-path` not being set up properly at the time of
`batch-byte-compile` execution.


[log.2017-04-01_11-56-03.wilddog.sh_home_brentg_logs_sh.1ef9d39d-c837-4ba2-a13c-36f79c7f8cd8.sh.log.txt](https://github.com/jorgenschaefer/elpy/files/887895/log.2017-04-01_11-56-03.wilddog.sh_home_brentg_logs_sh.1ef9d39d-c837-4ba2-a13c-36f79c7f8cd8.sh.log.txt)

**Comment by jorgenschaefer** [comment]
> What you suggested works, but now I want to know what caused this.

Did you upgrade Emacs?

**Comment by bgoodr** [comment]
No I didn't upgrade Emacs.

**Comment by jorgenschaefer** [comment]
Then I'm afraid I have no idea. Emacs Lisp macros have some weird effects when the files are compiled. Could be some old macro expansion in the .elc files with new functions or something. :-/

**Comment by bgoodr** [comment]
Yep! Sounds likely. Not sure I/we/anyone can debug do root cause analysis
down to a reasonable level. I'll give it some thought, so you can close
this as you see fit. If I come up with any brilliant observations, I'll
reply back. :)

**Comment by jorgenschaefer** [comment]
Thank you :-)

**Comment by bgoodr** [comment]
I narrowed it down to `.emacs.d/elpa/company-20170315.1754/.emacs.d/elpa/company-20170315.1754/company-dabbrev.el` that, when its .elc is deleted, things work. I found this because bytecompiling it using a `emacs --eval` script fails to byte compile it because the `load-path` is not correct.

Do you know when are these .elc files are supposed to be created? Meaning,  when do we expect the `.el` files to be byte-compiled? Is this done on the melpa servers and then copied down into the `.emacs.d` subdirectory as is, or is it expected that the local emacs is to do that byte compilation? I would expect the latter given your question about "did you upgrade Emacs"?  

Thanks,
Brent

**Comment by bgoodr** [comment]
The other thing I did was just blow away my `.emacs.d` directory entirely and just had all of the packages reinstalled from scratch, and that fixed the problem as well. So I suspect that particular version of company-dabbrev.el just didn't byte-compile correctly in this version of emacs: `GNU Emacs 24.5.1`

**Comment by jorgenschaefer** [comment]
.el files are compiled on package installation. My guess is that company-dabbrev uses some macro that is provided by some other package FOO. FOO-buggy was installed, company got installed, company-dabbrev was compiled with a buggy version of the macro from FOO. Then FOO fixed the bug and got upgraded, but the company-dabbrev .elc files remained. The next upgrade of company would have fixed things.

But that's just a guess.

**Comment by bgoodr** [comment]
Sounds reasonable to me. I admit I'm rather new to the whole package.el
thing. I don't even know what the proper way is to upgrade it, or the
packages it manages. I'll have to go web searching for the manual and read
it thoroughly.  I eventually will want to use it more and more now that it
seems to be much much more functional than it was years ago when I first
tried it and became frustrated with it back then. :)

**Comment by jorgenschaefer** [comment]
The proper way is to just hit `U x` in the `M-x list-packages` buffer. That this can break is more of a problem in how Emacs Lisp works (macros and byte compilation) than package.el, and there's not much you can do about it, sadly. The same problem could have happened if you installed packages manually.

**Comment by bgoodr** [comment]
Got it.
